rust/lance-table/src/rowids.rs

```diff
@@ -206,6 +212,215 @@ impl RowIdSequence {
+                return segment.get(index - offset);
+            }
+            offset += segment_len;
+        }
+        None
+    }
+}
+
+#[derive(Debug)]
+pub struct RowIdSeqSlice<'a> {
+    /// Current slice of the segments we cover
+    segments: &'a [U64Segment],
+    /// Offset into the first segment to start iterating from
+    offset_start: usize,
+    /// Offset into the last segment to stop iterating at
+    offset_last: usize,
+}
+
+impl<'a> RowIdSeqSlice<'a> {
+    pub fn iter(&self) -> impl Iterator<Item = u64> + '_ {
+        let mut known_size = self.segments.iter().map(|segment| segment.len()).sum();
+        known_size -= self.offset_start;
+        known_size -= self.segments.last().unwrap().len() - self.offset_last;
+
+        let end = self.segments.len() - 1;
+        self.segments
+            .iter()
+            .enumerate()
+            .flat_map(move |(i, segment)| {
+                match i {
+                    0 if self.segments.len() == 1 => {
+                        let len = self.offset_last - self.offset_start;
+                        // TODO: Optimize this so we don't have to use skip
+                        // (take is probably fine though.)
+                        Box::new(segment.iter().skip(self.offset_start).take(len))
+                            as Box<dyn Iterator<Item = u64>>
+                    }
+                    0 => Box::new(segment.iter().skip(self.offset_start)),
+                    i if i == end => Box::new(segment.iter().take(self.offset_last)),
+                    _ => Box::new(segment.iter()),
+                }
+            })
+            .exact_size(known_size)
+    }
+}
+
+/// Re-chunk a sequences of row ids into chunks of a given size.
+///
+/// # Errors
+///
+/// Will return an error if the sum of the chunk sizes is not equal to the total
+/// number of row ids in the sequences.
+pub fn rechunk_sequences(
+    sequences: impl IntoIterator<Item = RowIdSequence>,
+    chunk_sizes: impl IntoIterator<Item = u64>,
+) -> Result<Vec<RowIdSequence>> {
+    // TODO: return an iterator. (with a good size hint?)
+    let chunk_size_iter = chunk_sizes.into_iter();
+    let mut chunked_sequences = Vec::with_capacity(chunk_size_iter.size_hint().0);
+    let mut segment_iter = sequences
+        .into_iter()
+        .flat_map(|sequence| sequence.0.into_iter())
+        .peekable();
+
+    let mut segment_offset = 0_u64;
+    for chunk_size in chunk_size_iter {
+        let mut sequence = RowIdSequence(Vec::new());
+        let mut remaining = chunk_size;
+
+        let too_many_segments_error = || {
+            Error::invalid_input(
+                "Got too many segments for the provided chunk lengths",
+                location!(),
+            )
+        };
+
+        while remaining > 0 {
+            let remaining_in_segment = segment_iter
+                .peek()
+                .map_or(0, |segment| segment.len() as u64 - segment_offset);
+            use std::cmp::Ordering::*;
```

Comment:
Minor nit, maybe move this to top of file (also, I think you only need `Greater`)

But it's probably fine like this too.